File: app/ionic/src/server/Interfaces/Prayers.ts
// //
// //  Prayers.swift
// //  ordo-1962
// //
// //  Created by Matthew Frankland on 15/12/2023.
// //

// import Foundation
// import SwiftData

// @Model class PrayerLanguageData: Decodable {
//     private var res: [ String : PrayerCategoryData ] = [ : ]
    
//     required init ( from decoder: Decoder ) throws {
//         let values = try decoder.container ( keyedBy: CodingKeys.self )
//         self.res = [
//             "English": try values.decode ( PrayerCategoryData.self, forKey: .English ),
//             "Latin": try values.decode ( PrayerCategoryData.self, forKey: .Latin )
//         ]
//     }
    
//     private enum CodingKeys: String, CodingKey {
//         case English, Latin
//     }
    
//     func GetPrayer ( lang: String, category: String, name: String ) -> String {
//         if let language = res [ lang ] {
//             if let cat = language [ category ] {
//                 if let prayer = cat [ name ] {
//                     return prayer
//                 }
//             }
//         }
//         return ""
//     }
    
//     func GetPrayerNames ( lang: String, cat: String ) -> [ String ] {
//         if let lang = self.res [ lang ] {
//             if let prayer = lang [ cat ] {
//                 return prayer.keys.sorted ( by: < )
//             }
//         }
//         return [ ]
//     }
    
//     func GetCategories ( lang: String ) -> [ String ] {
//         if let lang = self.res [ lang ] {
//             return lang.keys.sorted ( by: < )
//         }
//         return [ ]
//     }
    
//     func GetLanguageDetails ( ) -> [ String ] {
//         return [ "English", "Latin" ].sorted ( by: < )
//     }
// }

// typealias PrayerCategoryData = [ String : PrayerData ]
// typealias PrayerData = [ String : String ]

File: app/android-studio/app/src/main/java/Server/Models/Prayers.java
package Server.Models;

import com.google.gson.annotations.SerializedName;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Prayers {
    private Map<String, PrayerCategoryData> res = new HashMap<>();

    public Prayers ( PrayerCategoryData english, PrayerCategoryData latin ) {
        res.put ( "English", english );
        res.put ( "Latin", latin );
    }

    public String getPrayer(String lang, String category, String name) {
        PrayerCategoryData categoryMap = this.res.get(lang);
        if (categoryMap != null) {
            PrayerData prayerMap = categoryMap.get(category);
            if (prayerMap != null) {
                return prayerMap.get(name);
            }
        }
        return "";
    }

    public List<String> getPrayerNames(String lang, String cat) {
        PrayerCategoryData categoryMap = this.res.get(lang);
        if (categoryMap != null) {
            return categoryMap.get(cat) != null ? categoryMap.get(cat).keySet().stream().sorted().toList() : List.of();
        }
        return List.of();
    }

    public List<String> getCategories(String lang) {
        PrayerCategoryData categoryMap = this.res.get(lang);
        return categoryMap != null ? categoryMap.keySet().stream().sorted().toList() : List.of();
    }

    public List<String> getLanguageDetails() {
        return List.of("English", "Latin").stream().sorted().toList();
    }
}

class PrayerCategoryData extends HashMap<String, PrayerData> { }

class PrayerData extends HashMap<String, String> { }
